dev/tools/lib/repository.dart

```diff
@@ -46,22 +46,34 @@ class Repository {
 
   Directory _checkoutDirectory;
 
-  /// Lazily-loaded directory for the repository checkout.
+  /// Directory for the repository checkout.
   ///
-  /// Cloning a repository is time-consuming, thus the repository is not cloned
-  /// until this getter is called.
+  /// Since cloning a repository takes a long time, we do not ensure it is
+  /// cloned on the filesystem until this getter is accessed.
   Directory get checkoutDirectory {
     if (_checkoutDirectory != null) {
       return _checkoutDirectory;
     }
     _checkoutDirectory = parentDirectory.childDirectory(name);
-    if (checkoutDirectory.existsSync() && !useExistingCheckout) {
-      deleteDirectory();
+    if (!useExistingCheckout && _checkoutDirectory.existsSync()) {
+      stdio.printTrace('Deleting $name from ${_checkoutDirectory.path}...');
+      _checkoutDirectory.deleteSync(recursive: true);
+    } else if (useExistingCheckout && _checkoutDirectory.existsSync()) {
+      git.run(
+        <String>['checkout', 'master'],
+        'Checkout to master branch',
+        workingDirectory: _checkoutDirectory.path,
+      );
+      git.run(
+        <String>['pull', '--ff-only'],
```

Comment:
Maybe bail here if master isn't tracking the upstream master branch (my setup is to track origin master)?  I have no idea if my setup is common, so also you can ignore and wait for anyone to actually hit that problem...

Reply:
1) I will apply some checks here, good idea;
2) This is currently just a debug feature, to keep me from thrashing my comcast quota while testing. I don't believe this flag is exposed in the tool's public API. By default, we always clone a fresh repo before operating on it, to ensure that the remotes are configured correctly.